implementation/src/main/java/io/smallrye/jwt/auth/principal/KeyLocationResolver.java

```diff
@@ -52,50 +53,56 @@
  * This implements the MP-JWT 1.1 mp.jwt.verify.publickey.location config property resolution logic
  */
 public class KeyLocationResolver implements VerificationKeyResolver {
-    private static final Logger log = Logger.getLogger(KeyLocationResolver.class);
+    private static final Logger LOGGER = Logger.getLogger(KeyLocationResolver.class);
 
     // The 'content' and 'httpsJwks' fields are used to keep the key content and mutually exclusive.
     // 'content' represents the key(s) loaded from all resources but the HTTPS URL based JWK set.
-    private String content;
+    private final String content;
     // 'httpsJwks' represents the JWK set loaded from the HTTPS URL.
-    private HttpsJwks httpsJwks;
+    private final HttpsJwks httpsJwks;
+    private final JWTAuthContextInfo authContextInfo;
 
     // If the 'smallrye.jwt.token.kid' is set then the verification key will be calculated
     // only once and used for all the token verification requests
     private volatile PublicKey verificationKey;
 
-    private JWTAuthContextInfo authContextInfo;
-
     public KeyLocationResolver(JWTAuthContextInfo authContextInfo) throws UnresolvableKeyException {
         this.authContextInfo = authContextInfo;
         try {
-            loadContents();
+            AbstractMap.SimpleImmutableEntry<HttpsJwks, String> contents = loadContents();
+            this.content = contents.getValue();
+            this.httpsJwks = contents.getKey();
         } catch (Exception e) {
             throw new UnresolvableKeyException("Failed to load a key from: " + authContextInfo.getPublicKeyLocation(), e);
         }
     }
 
     @Override
     public Key resolveKey(JsonWebSignature jws, List<JsonWebStructure> nestingContext) throws UnresolvableKeyException {
-        if (verificationKey != null) {
-            return verificationKey;
-        }
-        PublicKey key = tryAsJWK(jws);
-        if (key == null) {
-            key = tryAsPEM();
-        }
-        if (key == null) {
-            throw new UnresolvableKeyException("Failed to read location as any of JWK, JWKS, PEM. "
-                    + authContextInfo.getPublicKeyLocation());
-        }
-        return key;
-    }
-
-    private PublicKey tryAsPEM() {
-        try {
-            verificationKey = KeyUtils.decodePublicKey(content);
-        } catch (Exception e) {
-            log.debug("Failed to read location as PEM", e);
+        // httpsJwks may refresh itself even if the jwksRefreshInterval is set to 0 as it checks HTTPS Cache-Control header
+        // so the public key will have to be created per request to ensure the rotation (if any) is effective.
+        if (httpsJwks != null && authContextInfo.getTokenKeyId() == null)
+            return tryAsJWK(jws);
```

Comment:
I think in general this is better cleanup, the initial attempt to consolidate the key processing inside the resolver was not perfect.  But I'd like to avoid calling `tryAsJwk` twice here since there will again some redundant checks done there. 
I propose to remove `tryAsJwk` completely, but move its code extracting the `kid` into into a `getKid(JsonWebSignature)` method (I will need it anyway for #85 as well to match the kid against the key alias).
Next here a call to tryAsHttpsJwk(jws) will be made, which will get the kid from `getKid(Jws)`